xgcm/autogenerate.py

```diff
@@ -0,0 +1,254 @@
+                The direction in which to shift the array. If not specified,
+                default will be used.
+            boundary : {None, 'fill', 'extend'}
+                A flag indicating how to handle boundaries:
+                * None:  Do not apply any boundary conditions. Raise an error
+                if boundary conditions are required for the operation.
+                * 'fill':  Set values outside the array boundary to fill_value
+                  (i.e. a Neumann boundary condition.)
+                * 'extend': Set values outside the array to the nearest array
+                  value. (i.e. a limited form of Dirichlet boundary condition.)
+
+            fill_value : float, optional
+                 The value to use in the boundary condition with
+                 `boundary='fill'`.
+
+            wrap_value : float, optional
+                if value is given, cyclic coordinate is wrapped approriately,
+                e.g. longitude is wrapped around 180 or 360 discontinuity
+            Returns
+            -------
+            da_i : xarray.DataArray
+                The differenced data
+            """
+            #TODO I think we can modify xgcm.Axis so that this can all be
+            # absorbed in there.
+
+            position_from, dim = self._get_axis_coord(da)
+            if to is None:
+                to = self._default_shifts[position_from]
+
+            # get the two neighboring sets of raw data
+            data_left, data_right = \
+                self._get_neighbor_data_pairs(da, to,
+                                              boundary=boundary,
+                                              fill_value=fill_value)
+
+            # Is there a way to use the existing grid.Axis.interp logic?
+            # For now just hardcode it
+            data_new = 0.5*(data_left + data_right)
+
+            return data_new
+
+
+def generate(ds, axes, dims, to,
+             x_wrap, y_wrap, z_pad,
+             to_shift, from_shift, dim_switch=True):
+        """
+        Generate new dataarray from coordinates or dimensions, which might be
+        missing in order to get a 'full' c-grid.
+
+        INPUTS:
+
+        ds - xarray.Dataset
+
+        axes = dictionary containing comodo axis and corresponding DataArray
+        in ds (e.g. axes={'X':'llon','Y':'llon'}). These could be
+        multidimensional coordinates like geospatial longitudes/latitudes
+
+        dims = dictionary of dimensons in ds corresponding to comodo axis
+        e.g. (dims={'X':'lon','Y':'lon'}). These have to be names of
+        strictly dimensions of ds.
+
+        to - {'left','right'} Direction of shift, similar as the logic in
+        xgcm.Axis, needed to apply shift correctly
+
+        x_wrap - float: value of discontinuity in x direction (defaults to
+        360 for global longitude)
+
+        y_wrap - float: as x_wrap but for y direction (defaults to
+        180 for global longitude)
```

Comment:
My intention was to make this work relatively smoothly "out of the box with common datasets (like satellite products). The logic should work with any succession of values that need to be wrapped around a discontinuity. This would have to be explained better in the docstring, but if you have an arbitrary dataset with x coordinates 1,2,...10 you would have to specify 'x_wrap=10' and it should do the right thing. Should have a test too.